python/cuml/common/array_sparse.py

```diff
@@ -0,0 +1,148 @@
+    shape : tuple of ints
+        Shape of the array
+    nnz : int
+        Number of nonzeros in underlying arrays
+    has_sorted_indices : bool
+        Whether column indices and data are sorted by column
+    """
+
+    @with_cupy_rmm
+    def __init__(self, data=None, dtype=None):
+        if not cpx.scipy.sparse.isspmatrix(data) and \
+                not (has_scipy() and scipy.sparse.isspmatrix(data)):
+            raise ValueError("A sparse matrix is expected as input. "
+                             "Received %s" % type(data))
+
+        data = data.tocsr()  # currently only CSR is supported
+
+        # Note: Only 32-bit indexing is supported currently.
+        # In CUDA11, Cusparse provides 64-bit function calls
+        # but these are not yet used in RAFT/Cuml
+        self.indptr, _, _, _ = input_to_cuml_array(data.indptr,
+                                                   check_dtype=False,
+                                                   convert_to_dtype=np.int32)
+
+        self.indices, _, _, _ = input_to_cuml_array(data.indices,
+                                                    check_dtype=False,
+                                                    convert_to_dtype=np.int32)
+
+        self.data, _, _, _ = input_to_cuml_array(data.data,
+                                                 check_dtype=False,
+                                                 convert_to_dtype=dtype)
+
+        self.shape = data.shape
+        self.dtype = self.data.dtype
+        self.nnz = data.nnz
+        self.has_sorted_indices = data.has_sorted_indices
+
+    @with_cupy_rmm
+    def to_output(self, output_type='cupy',
+                  output_format=None,
+                  output_dtype=None):
+        """
+        Convert array to output format
+
+        Parameters
+        ----------
+        output_type : string
+            Format to convert the array to. Acceptable formats are:
+
+            - 'cupy' - to cupy array
+            - 'scipy' - to scipy (host) array
+
+        output_format : string, optional { 'coo', 'csc' }
+            Optionally convert the output to the specified format.
+        output_dtype : string, optional
+            Optionally cast the array to a specified dtype, creating
+            a copy if necessary.
+        """
+        output_dtype = self.data.dtype \
+            if output_dtype is None else output_dtype
+
+        if output_type not in ['cupy', 'scipy']:
+            raise ValueError("Unsupported output_type: %s" % output_dtype)
+
+        cuml_arr_output_type = 'numpy' if output_type == 'scipy' else 'cupy'
+
+        data = self.data.to_output(cuml_arr_output_type, output_dtype)
+        indices = self.indices.to_output(cuml_arr_output_type)
+        indptr = self.indptr.to_output(cuml_arr_output_type)
+
+        if output_type == 'cupy':
+            constructor = cpx.scipy.sparse.csr_matrix
+
+        elif output_type == 'scipy' and has_scipy():
+            if has_scipy():
```

Comment:
Future/separate change:
 Should we change our import_utils functions to have a "raise_if_unavailable" input argument that does this? Seems like a pretty common pattern - raise a friendly exception if not here. Or maybe it's not that common and we can just get the functionality by doing `import foo`...

Reply:
Yeah, I think that would be very useful. I'll add it in this PR.